test-harness/src/matcher.rs

```diff
@@ -0,0 +1,223 @@
+    pub fn span_close() -> Self {
+        // This regex will never fail to parse.
+        Self::message("close").unwrap()
+    }
+
+    /// Require that matching events be in a span with the given name.
+    ///
+    /// Note that this will overwrite any previously-set spans.
+    pub fn in_span(mut self, span: &str) -> Self {
+        self.spans.clear();
+        self.spans.push(span.to_owned());
+        self
+    }
+
+    /// Require that matching events be in spans with the given names.
+    ///
+    /// Spans are listed from the inside out; that is, a call to `in_spans(["a", "b", "c"])` will
+    /// require that events be emitted from a span `a` directly nested in a span
+    /// `b` directly nested in a span `c`.
+    ///
+    /// All listed spans must be present in the correct order, but do not otherwise need to be
+    /// "anchored" or uninterrupted.
+    ///
+    /// Note that this will overwrite any previously-set spans.
+    pub fn in_spans(mut self, spans: impl IntoIterator<Item = impl AsRef<str>>) -> Self {
+        self.spans = spans.into_iter().map(|s| s.as_ref().to_owned()).collect();
+        self
+    }
+
+    /// Require that matching events be emitted from the given module as recorded by the event's
+    /// `target` field.
+    ///
+    /// Note that this requires the module name to match exactly; child modules will not be
+    /// matched.
+    pub fn in_module(mut self, module: &str) -> Self {
+        self.target = Some(module.to_owned());
+        self
+    }
+
+    /// Determines if this query matches the given event.
+    pub fn matches(&self, event: &Event) -> bool {
+        if !self.message.is_match(&event.message) {
+            return false;
+        }
+
+        if !self.spans.is_empty() {
+            let mut spans = event.spans();
+            for expected_name in &self.spans {
+                loop {
+                    match spans.next() {
+                        Some(span) => {
+                            if &span.name == expected_name {
+                                // Found this expected span, move on to the next one.
+                                break;
+                            }
+                            // Otherwise, this span isn't the expected one, but the next span might
+                            // be.
+                        }
+                        None => {
+                            // We still expect to see another span, but there's no spans left in
+                            // the event.
+                            return false;
+                        }
+                    }
+                }
+            }
+        }
+
+        if let Some(target) = &self.target {
+            if target != &event.target {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+/// A type that can be converted into a `Matcher` and used for searching log events.
+pub trait IntoMatcher {
```

Comment:
i feel like this `IntoMatcher` class is probably unnecessary but I don't feel too strongly about it

my reasoning is that usually interfaces are good if you want to program abstractly against that interface (i.e. write a derived function that works for all possible `IntoMatcher`s), but the impression I get so far is that this interface is only used concretely (i.e. to instantiate a specific `Matcher` directly).  In those cases it would probably be just as easy to use `Matcher::message` directly since the type of the thing you are wrapping is statically known in that case.

Reply:
Yeah, so the test harness (in the next PR) exposes a `get_log(matcher: impl IntoMatcher)` method so you can give it either a string literal (for simple cases) or a `Matcher` directly (for more complex cases).